Blockchain params endpoint - part 1

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#1291 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added new endpoint scaffolding with needed types
- [x] I have extended swagger
- [x] I have updated unit tests


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
introduce GetNetworkInformation

paweljakubas committed:
add GetNetworkParameters and propagate it through codebase

paweljakubas committed:
add corresponding swagger part

paweljakubas committed:
update Api.TypesSpec and be more complient with requirement

Anviking commented on lib/core/src/Cardano/Wallet/Api.hs:
```diff
@@ -62,6 +62,8 @@ module Cardano.Wallet.Api
 
     -- * Miscellaneous
     , Network
+        , GetNetworkInformation
```
These two imports look misaligned

paweljakubas replied:
just followed pattern from the above for other endpoints

Anviking replied:
Ah, I see 👍

Anviking commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -446,6 +466,33 @@ instance FromHttpApiData Iso8601Time where
 
+instance FromText ApiEpochNumber where
+    fromText txt = case txt of
+        "latest" -> Right (ApiEpochNumber "latest")
```
🤔 What about defining `ApiEpochNumber` as something like:

```haskell
data ApiEpochNumber = ApiEpochNumberLatest | ApiEpochNumber EpochNumber
```

Anviking replied:
It seems like when actually using `ApiEpochNumber`, you'd have to re-parse the `Text` into either the notion of latest or an `EpochNumber` anyway.

paweljakubas replied:
I was persuaded and used the propose type. It is more precise and better reflect situation

Anviking commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -787,6 +800,25 @@ spec = do
+                            activeSlotCoefficient (x :: ApiNetworkParameters)
+                    }
+            in
+                x' === x .&&. show x' === show x
```
> "pointless tests to trigger coverage for record accessors"

Hm, ok 👍 

Unrelated to the PR, but do we still care about coverage to this extent? We have removed the badge from the readme at least.

paweljakubas replied:
yeah, `pointless tests` title says it all

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1306,6 +1310,14 @@ getNetworkInformation (_block0, bp, st) nl = do
+    -> NetworkLayer IO t Block
+    -> ApiEpochNumber
+    -> Handler ApiNetworkParameters
+getNetworkParameters _ _ _ = error "not implemented"
```
I think best practice is to `throwError err501` instead of erroring.

paweljakubas replied:
I would do this if we had not `Handler` but `ExceptT` . When `Handler` then I need to do : 
```
liftHandler $ throwE err501
```
and then we have 
```
No instance for (LiftHandler ServantErr) 
```
So I would rather to stick to error and in replace it in the next PR with implementation

paweljakubas replied:
ok, but we can use here : 
```
liftIO $ throwIO err501
```

Anviking replied:
I have a memory of `err501` being used in the wallet, but I don't remember exactly how.

With `liftIO $ throwIO err501` we actually get a 500:
```
http http://localhost:8090/v2/network/parameters/latest
HTTP/1.1 500 Internal Server Error
Content-Type: text/plain; charset=utf-8
Date: Thu, 23 Jan 2020 10:08:45 GMT
Server: Warp/3.2.27
Transfer-Encoding: chunked

Something went wrong
```

Which is the same result as `error "not implemeted"`.

Whereas with `Handler $ throwE err501` we at least get a 501 (but with generic code and message)
```
http http://localhost:8090/v2/network/parameters/latest
HTTP/1.1 501 Not Implemented
Content-Type: application/json;charset=utf-8
Date: Thu, 23 Jan 2020 10:10:33 GMT
Server: Warp/3.2.27
Transfer-Encoding: chunked

{
    "code": "unexpected_error",
    "message": "It looks like something unexpected went wrong. Unfortunately I don't yet know how to handle this type of situation. Here's some information about what happened: "
}
```

paweljakubas replied:
agree. used `Handler $ throwE err501`

paweljakubas replied:
in the end used Servant's `throwError err501`

Anviking commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1171,6 +1203,31 @@ instance Arbitrary ApiNetworkInformation where
+instance Arbitrary (Quantity "second" NominalDiffTime) where
+    shrink (Quantity 0.0) = []
+    shrink _ = [Quantity 0.0]
+    arbitrary = Quantity . fromInteger <$> choose (0, 100)
```
Seems ok, but why so small?

paweljakubas replied:
ok, increased to 10000

Anviking commented on specifications/api/swagger.yaml:
```diff
@@ -1732,12 +1800,23 @@ paths:
+      parameters:
+        - *parametersEpoch
+      description: |
+        <p align="right">status: <strong>under development</strong></p>
```
🆒 👍

Anviking reviewed (changes requested):
Looks good! Left a few comments. Most notably I think `ApiEpochNumber` could be less stringly typed.

paweljakubas committed:
improve on ApiEpochNumber type and other fixes

clean arbitrary

paweljakubas committed:
change epoch -> epochId in swagger.yaml

fix attempt

use better response for unimplemented API endpoint

epoch -> epochId

fix swagger

use throwError from Servant to service err501

update Wallet.Api literal in endpoint

Anviking reviewed (approved):
lgtm

Anviking commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -334,6 +341,18 @@ newtype ApiFee = ApiFee
+    , slotLength :: !(Quantity "second" NominalDiffTime)
+    , epochLength :: !(Quantity "slot" Word32)
+    , epochStability :: !(Quantity "block" Word32)
+    , activeSlotCoefficient :: !(Quantity "percent" Percentage)
```
> Percentage

Though I think you might find in the next PR that `ActiveSlotCoeff` is a `Double` whereas `Percentage` is a `Word, [0, 100]`

Anviking replied:
Ok, U/S says:
> active_slot_coefficient: *percentage

And in `x-percentage` is a `number`. So it should be able to be a floating point.

I imagine some users would like to have the full `33.333333%` instead of just `33%`.
